quinn/src/lib.rs

```diff
@@ -879,16 +873,34 @@ impl Future for Driver {
             let mut fired = false;
             loop {
                 match endpoint.timers.poll() {
-                    Ok(Async::Ready(Some(Some((conn, timer))))) => {
+                    Ok(Async::Ready(Some(expired))) => {
+                        let (conn, timer) = expired.into_inner();
                         trace!(endpoint.log, "timeout"; "timer" => ?timer);
                         endpoint.inner.timeout(now, conn, timer);
                         fired = true;
                     }
-                    Ok(Async::Ready(Some(None))) => {}
+                    // Ready(None) is returned if there's not active timer, and NotReady is
+                    // returned if there are active timers but none expired yet.
                     Ok(Async::Ready(None)) | Ok(Async::NotReady) => {
                         break;
                     }
-                    Err(()) => unreachable!(),
+                    Err(timer_err) => {
+                        if timer_err.is_shutdown() {
+                            // This miht only happen is there's a bug in some low level tokio
+                            // components.
```

Comment:
I'm not sure this comment is strictly correct; it might be possible to produce this behavior by misusing low-level public tokio APIs as well. We should be careful when pointing fingers. Also, typo on "might".